File: backspace-string-compare/shaun-gy-2p.java
//Runtime: 0 ms, faster than 100.00% of Java online submissions for Backspace String Compare.
//shaun off gy idea 2p from back
class Solution {
    public boolean backspaceCompare(String S, String T) {
        int i = S.length()-1, j = T.length()-1;
        int count1 = 0, count2 = 0; //count of # seen from back
        while (i >= 0 || j >= 0) {
            
            //if current char is # or seen # from prev
            while (i >= 0 && (S.charAt(i) == '#' || count1 > 0 )) {
                if (S.charAt(i) == '#') count1++;
                else count1--;
                i--;
            }
            while (j >= 0 && (T.charAt(j) == '#' || count2 > 0 )) {
                if (T.charAt(j) == '#') count2++;
                else count2--;
                j--;
            }
            if (i < 0 || j < 0) return i == j;
            if (S.charAt(i) != T.charAt(j)) return false;
            i--;
            j--;
        }
        return i==j;
    }
}

File: backspace-string-compare/shaun-gy-2p.py
#Runtime: 28 ms, faster than 70.88% of Python3 online submissions for Backspace String Compare.
class Solution:
    def backspaceCompare(self, S: str, T: str) -> bool:
        i,j = len(S)-1, len(T)-1  #bug len --> len -1
        count1,count2 = 0,0
        while (i >= 0 or j >= 0):
            while (i >=0 and (S[i] == '#' or count1 > 0)):
                if S[i] == '#': count1+=1
                else: count1-=1
                i-=1
            
            while (j >=0 and (T[j] == '#' or count2 > 0)):
                if T[j] == '#': count2+=1
                else: count2-=1
                j-=1
            if (i < 0 or j < 0): return i==j
            if S[i] != T[j]: return False
            i-=1
            j-=1
        return i==j

